numpy/lib/function_base.py

```diff
@@ -4305,80 +4441,148 @@ def _percentile(a, q, axis=None, out=None,
     if axis is None:
         axis = 0
 
-    Nx = ap.shape[axis]
-    indices = q * (Nx - 1)
+    # move axis to -1 for np.vectorize() operations.  Will move back later.
+    ap = np.swapaxes(ap, axis, -1)
+
+    def _sort_by_index(vector, vec_indices):
+        return vector[vec_indices]
+
+    # this func vectorizes sort along axis
+    arraysort = np.vectorize(_sort_by_index, signature='(i),(i)->(i)')
+
+    if weights is None:
+        Nx = ap.shape[-1]
+        indices = q * (Nx - 1)
+        indices = np.broadcast_to(indices, ap.shape[:-1] + q.shape)
+        ind_sorted = np.argsort(ap, axis=-1)  # sort values long axis
+        ap = arraysort(ap, ind_sorted)
```

Comment:
without weights it should not be slower than the old percentile, so partition should be used and the vectorization of the index sort avoided.